src/sources/journald.rs

```diff
@@ -0,0 +1,281 @@
+use crate::{
+    event,
+    event::{Event, LogEvent, ValueKind},
+    topology::config::{DataType, GlobalOptions, SourceConfig},
+};
+use chrono::TimeZone;
+use futures::{future, sync::mpsc, Future, Sink};
+use journald::{Journal, Record};
+use lazy_static::lazy_static;
+use serde::{Deserialize, Serialize};
+use std::collections::HashSet;
+use std::io::Error;
+use std::sync::mpsc::RecvTimeoutError;
+use std::thread;
+use std::time;
+use string_cache::DefaultAtom as Atom;
+use tracing::{dispatcher, field};
+
+lazy_static! {
+    static ref MESSAGE: Atom = Atom::from("MESSAGE");
+    static ref TIMESTAMP: Atom = Atom::from("_SOURCE_REALTIME_TIMESTAMP");
+    static ref HOSTNAME: Atom = Atom::from("_HOSTNAME");
+}
+
+#[derive(Deserialize, Serialize, Debug, Default)]
+#[serde(deny_unknown_fields, default)]
+pub struct JournaldConfig {
+    pub current_runtime_only: Option<bool>,
+    pub local_only: Option<bool>,
+    pub units: Vec<String>,
+}
+
+fn map_bool<T>(flag: bool, t_value: T, f_value: T) -> T {
+    match flag {
+        true => t_value,
+        false => f_value,
+    }
+}
+
+#[typetag::serde(name = "journald")]
+impl SourceConfig for JournaldConfig {
+    fn build(
+        &self,
+        _name: &str,
+        _globals: &GlobalOptions,
+        out: mpsc::Sender<Event>,
+    ) -> Result<super::Source, String> {
+        let flags = map_bool(
+            self.current_runtime_only.unwrap_or(true),
+            journald::SD_JOURNAL_RUNTIME_ONLY,
+            0,
+        ) | map_bool(
+            self.local_only.unwrap_or(true),
+            journald::SD_JOURNAL_LOCAL_ONLY,
+            0,
+        );
+        let journal = Journal::open(flags).map_err(|err| format!("{}", err))?;
```

Comment:
from the errors being produced in the journald lib, it looks like they may not be descriptive enough? We probably want to link them out to way to install the libsystemd etc. What do you think?

Reply:
I'm not clear what you're asking about here.

Reply:
Sorry, trying to think about how we might communicate to the user that the libsystemd was not found. I curious if it may be worth it to write a custom error message here that is a bit descriptive since not finding the lib can be a pain and confusing. Just thinking about the users UX.

Reply:
I think this has been dealt with — the new error return type from `Journal::open` distinguishes between an error in `dlopen` and another I/O error while opening the journal itself, which will in turn be reflected in the error message.